mlflow/models/evaluation/default_evaluator.py

```diff
@@ -258,6 +261,59 @@ def _do_plot(**kwargs):
 }
 
 
+def _evaluate_custom_metric(index, custom_metric, eval_df, builtin_metrics):
+    """
+    This function calls the `custom_metric` function and performs validations on the returned
+    result to ensure that they are in the expected format. It will raise a MlflowException if
+    the result is not in the expected format.
+
+    :param index
+    :param custom_metric: A user provided custom metric function
+    :param eval_df: A Pandas dataframe object containing a prediction and a target column
+    :param builtin_metrics: A dictionary of metrics produced by the default evaluator
+    :return: A tuple of dictionaries. The first is a dictionary of metrics, the second is
+             a dictionary of artifacts (which can be null if the custom metric function did
+             not produce any).
+    """
+    result = custom_metric(eval_df, builtin_metrics)
+    if result is None:
+        _logger.warning(
+            f"Custom metric function "
+            f"'{getattr(custom_metric, '__name__', repr(custom_metric))}' at index "
+            f"{index} in the `custom_metrics` parameter returned None."
+        )
+        return None, None
+
+    def __validate_metrics_dict(metrics):
+        return isinstance(metrics, dict) and all(
+            isinstance(metric_name, str) and isinstance(metric_val, (int, float, np.number))
+            for metric_name, metric_val in metrics.items()
+        )
+
+    def __validate_artifacts_dict(artifacts):
+        return isinstance(artifacts, dict) and all(
+            isinstance(artifacts_name, str) for artifacts_name in artifacts.keys()
```

Comment:
If any of these condition is not true, raise in place with proper error message.